Code/GraphMol/RascalMCES/RascalMCES.cpp

```diff
@@ -0,0 +1,1086 @@
+  std::vector<size_t> a(std::min(atomDegrees1.size(), atomDegrees2.size()),
+                        99999999);
+  std::vector<size_t> b(std::min(atomDegrees1.size(), atomDegrees2.size()),
+                        99999999);
+  int retVal = lap_maximize(costsMat, a, b);
+  if (retVal < 0) {
+    // no solution for the LAP was possible.
+    return 0;
+  }
+  int totalCost = 0;
+  for (auto i = 0u; i < a.size(); ++i) {
+    totalCost += costsMat[a[i]][b[i]];
+  }
+  return totalCost;
+}
+
+namespace details {
+// calculate the tier 1 similarity between the 2 mols.
+double tier1Sim(const ROMol &mol1, const ROMol &mol2,
+                std::map<int, std::vector<std::pair<int, int>>> &degSeqs1,
+                std::map<int, std::vector<std::pair<int, int>>> &degSeqs2) {
+  sorted_degree_seqs(mol1, degSeqs1);
+  sorted_degree_seqs(mol2, degSeqs2);
+  int vg1g2 = 0;
+  int eg1g2 = 0;
+  for (const auto &it1 : degSeqs1) {
+    const auto &seq2 = degSeqs2.find(it1.first);
+    if (seq2 != degSeqs2.end()) {
+      vg1g2 += std::min(it1.second.size(), seq2->second.size());
+      auto numToDo = std::min(it1.second.size(), seq2->second.size());
+      for (auto i = 0U; i < numToDo; ++i) {
+        eg1g2 += std::min(it1.second[i].first, seq2->second[i].first);
+      }
+    }
+  }
+  eg1g2 /= 2;
+  double sim = double((vg1g2 + eg1g2) * (vg1g2 + eg1g2)) /
+               double((mol1.getNumAtoms() + mol1.getNumBonds()) *
+                      (mol2.getNumAtoms() + mol2.getNumBonds()));
+  return sim;
+}
+// Calculate the tier 2 similarity between the 2 mols.
+double tier2Sim(const ROMol &mol1, const ROMol &mol2,
+                const std::map<int, std::vector<std::pair<int, int>>> &degSeqs1,
+                const std::map<int, std::vector<std::pair<int, int>>> &degSeqs2,
+                const std::vector<unsigned int> &bondLabels1,
+                const std::vector<unsigned int> &bondLabels2) {
+  int vg1g2 = 0;
+  int eg1g2 = 0;
+  for (const auto &it1 : degSeqs1) {
+    const auto &seq2 = degSeqs2.find(it1.first);
+    if (seq2 != degSeqs2.end()) {
+      vg1g2 += std::min(it1.second.size(), seq2->second.size());
+      eg1g2 += getAssignmentScore(it1.second, seq2->second, bondLabels1,
+                                  bondLabels2, mol1, mol2);
+    }
+  }
+  eg1g2 /= 2;
+  double sim = double((vg1g2 + eg1g2) * (vg1g2 + eg1g2)) /
+               double((mol1.getNumAtoms() + mol1.getNumBonds()) *
+                      (mol2.getNumAtoms() + mol2.getNumBonds()));
+  return sim;
+}
+// For each bond in the molecule, encode it with its type and the labels of the
+// two end atoms, returning the results as strings.
+// Note that the molecule should not be in kekulized form.
+void getBondLabels(const ROMol &mol, const RascalOptions &opts,
+                   std::vector<std::string> &bondLabels) {
+  std::vector<std::string> atomLabels;
+  get_atom_labels(mol, opts, atomLabels);
+  bondLabels = std::vector<std::string>(mol.getNumBonds(), "");
```

Comment:
It might be picky, but I read Herb Sutter or Scott Meyers or someone once who said that writing code that shows you understand default values and the like gives the reader faith that you know what you're doing, so is no bad thing.